query/upgrade_test.go

```diff
@@ -27,36 +28,70 @@ import (
 )
 
 func TestMain(m *testing.M) {
-	conf := dgraphtest.NewClusterConfig().WithNumAlphas(3).WithNumZeros(3).WithReplicas(3).
-		WithACL(20 * time.Second).WithEncryption().WithVersion("0c9f60156")
-	c, err := dgraphtest.NewLocalCluster(conf)
-	x.Panic(err)
-	defer c.Cleanup()
-	c.Start()
-
-	// setup the global Cluster var
-	dc = c
-
-	// setup client
-	dg, err := c.Client()
-	if err != nil {
-		panic(err)
+	mutate := func(c dgraphtest.Cluster) {
+		dg, err := c.Client()
+		x.Panic(err)
+
+		client = dg
+		dc = c
+		populateCluster()
 	}
-	client = dg
 
-	// do mutations
-	populateCluster()
+	query := func(c dgraphtest.Cluster) {
+		dg, err := c.Client()
+		x.Panic(err)
+
+		client = dg
+		dc = c
+		if m.Run() != 0 {
+			panic("tests failed")
+		}
+	}
 
-	// upgrade
-	x.Panic(c.Upgrade("v23.0.0-beta1"))
+	runTest := func(before, after string) {
+		conf := dgraphtest.NewClusterConfig().WithNumAlphas(1).WithNumZeros(1).
+			WithReplicas(1).WithACL(time.Hour).WithVersion(before)
+		c, err := dgraphtest.NewLocalCluster(conf)
+		x.Panic(err)
+		defer c.Cleanup()
+		c.Start()
 
-	// setup the client again
-	dg, err = c.Client()
-	if err != nil {
-		panic(err)
+		mutate(c)
+		x.Panic(c.Upgrade(after, dgraphtest.BackupRestore))
+		query(c)
 	}
-	client = dg
 
-	// Run tests
-	_ = m.Run()
+	comboVersions := []struct {
+		before, after string
+	}{
+		// {"v20.11.3", "v23.0.0-beta1"},
+		// {"v21.03.0", "v23.0.0-beta1"},
+		{"v21.03.0-92-g0c9f60156", "v23.0.0-beta1"},
+		// {"v21.03.0-96-g65fff46c4-slash", "v23.0.0-beta1"},
+		// {"v21.03.0-98-g19f71a78a-slash", "v23.0.0-beta1"},
+		// {"v21.03.0-99-g4a03c144a-slash", "v23.0.0-beta1"},
+		// {"v21.03.1", "v23.0.0-beta1"},
+		// {"v21.03.2", "v23.0.0-beta1"},
+		// {"v21.12.0", "v23.0.0-beta1"},
+		// {"v22.0.0", "v23.0.0-beta1"},
+		{"v22.0.1", "v23.0.0-beta1"},
+		// {"v22.0.2", "v23.0.0-beta1"},
```

Comment:
we might want to paralellize this in the future, perhaps by making this configurable via a command line argument to `go test`

Reply:
We can just run the official go test using m.Run("name"). If we use proper names, go test will run a specific test only.

Reply:
no, that works when we are inside a Test not in TestMain